File: crates/aiperf-graph/src/channel_store.rs
// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0
//! Per-trace channel state for the async-dataflow executor.
//!
//! A versioned append-only log per channel; writes are linearized by a single
//! monotonic sequence counter (`last_seq`); readers capture per-channel versions and reducers
//! consume them in `(write_seq, writer_node_id)` order. Values are stored by
//! clone of the `serde_json::Value`, never mutated in place.

use crate::model::{ChannelSpec, Count, ReducerName};
use crate::reducers::{ChanVal, ReducerError, apply_reducer};
use serde_json::Value;
use std::cell::RefCell;
use std::cmp::Ordering;
use std::collections::BTreeMap;
use std::rc::Rc;
use tokio::sync::Notify;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum StoreError {
    UnknownChannel(String),
    Orphaned { channel: String, reason: String },
    Reducer(ReducerError),
}

impl std::fmt::Display for StoreError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            StoreError::UnknownChannel(c) => write!(f, "unknown channel: {c:?}"),
            StoreError::Orphaned { channel, reason } => {
                write!(f, "channel {channel:?} orphaned: {reason}")
            }
            StoreError::Reducer(e) => write!(f, "{e}"),
        }
    }
}
impl std::error::Error for StoreError {}

impl From<ReducerError> for StoreError {
    fn from(e: ReducerError) -> Self {
        StoreError::Reducer(e)
    }
}

/// One committed write to a value channel.
#[derive(Debug, Clone)]
struct LogEntry {
    write_seq: i64,
    writer_node_id: String,
    value: Value,
}

/// Canonical reducer-consumption order for two log entries: by `write_seq`, then
/// by `writer_node_id` as a stable tie-break for same-seq concurrent writes.
fn entry_order(a: &LogEntry, b: &LogEntry) -> Ordering {
    a.write_seq
        .cmp(&b.write_seq)
        .then_with(|| a.writer_node_id.cmp(&b.writer_node_id))
}

/// Immutable snapshot of which channel versions count for one firing.
#[derive(Debug, Clone, Default)]
pub struct VersionCapture {
    pub captured_seqs: Vec<i64>,
}

struct StoreInner {
    specs: BTreeMap<String, ChannelSpec>,
    logs: BTreeMap<String, Vec<LogEntry>>,
    arrival_count: BTreeMap<String, i64>,
    producers_remaining: BTreeMap<String, i64>,
    producers_declared: BTreeMap<String, i64>,
    /// The last committed write sequence number (0 when only init seeds exist).
    last_seq: i64,
    /// One `Notify` per channel; readers park on it and re-check their arrival
    /// count / orphan state on wake. `notify_waiters` wakes all parked readers
    /// of a channel (single-threaded, so a notify can't be lost between a
    /// reader's synchronous check and its `.await`).
    notifiers: BTreeMap<String, Rc<Notify>>,
    orphaned: BTreeMap<String, String>,
    overwrite_writer: BTreeMap<String, String>,
}

/// Per-trace channel store backed by per-channel append-only logs.
pub struct VersionedChannelStore {
    inner: RefCell<StoreInner>,
}

impl VersionedChannelStore {
    pub fn new(
        initial: &BTreeMap<String, Value>,
        channel_specs: &BTreeMap<String, ChannelSpec>,
        producers_per_channel: &BTreeMap<String, i64>,
    ) -> Result<Self, StoreError> {
        let specs = channel_specs.clone();
        let mut logs: BTreeMap<String, Vec<LogEntry>> = channel_specs
            .keys()
            .map(|c| (c.clone(), Vec::new()))
            .collect();
        let arrival_count: BTreeMap<String, i64> =
            channel_specs.keys().map(|c| (c.clone(), 0)).collect();
        let producers_remaining: BTreeMap<String, i64> = channel_specs
            .keys()
            .map(|c| (c.clone(), *producers_per_channel.get(c).unwrap_or(&0)))
            .collect();
        let producers_declared = producers_remaining.clone();
        let notifiers: BTreeMap<String, Rc<Notify>> = channel_specs
            .keys()
            .map(|c| (c.clone(), Rc::new(Notify::new())))
            .collect();

        // Seed initial state at write_seq 0 (reducer seed; not an arrival).
        for (ch, value) in initial {
            if !specs.contains_key(ch) {
                return Err(StoreError::UnknownChannel(ch.clone()));
            }
            logs.get_mut(ch).unwrap().push(LogEntry {
                write_seq: 0,
                writer_node_id: "__init__".to_string(),
                value: value.clone(),
            });
        }

        Ok(VersionedChannelStore {
            inner: RefCell::new(StoreInner {
                specs,
                logs,
                arrival_count,
                producers_remaining,
                producers_declared,
                last_seq: 0,
                notifiers,
                orphaned: BTreeMap::new(),
                overwrite_writer: BTreeMap::new(),
            }),
        })
    }

    // ---- write ----------------------------------------------------------

    /// Commit `value` to every channel in `channel_names`.
    pub fn write(
        &self,
        channel_names: &[String],
        value: &Value,
        writer_node_id: &str,
    ) -> Result<(), StoreError> {
        if channel_names.is_empty() {
            return Ok(());
        }
        {
            let inner = self.inner.borrow();
            for ch in channel_names {
                validate_write_channel(&inner, ch, writer_node_id)?;
            }
        }
        for ch in channel_names {
            {
                let mut inner = self.inner.borrow_mut();
                commit_write_channel(&mut inner, ch, value, writer_node_id);
            }
            self.notify_channel(ch);
        }
        Ok(())
    }

    // ---- await_inputs ---------------------------------------------------

    /// Block until every requirement is satisfied; return frozen captures.
    pub async fn await_inputs(
        &self,
        requirements: &[(String, Count)],
    ) -> Result<BTreeMap<String, VersionCapture>, StoreError> {
        let mut captures: BTreeMap<String, VersionCapture> = BTreeMap::new();
        for (channel, count) in requirements {
            {
                let inner = self.inner.borrow();
                if !inner.specs.contains_key(channel) {
                    return Err(StoreError::UnknownChannel(channel.clone()));
                }
                if let Some(reason) = inner.orphaned.get(channel) {
                    return Err(StoreError::Orphaned {
                        channel: channel.clone(),
                        reason: reason.clone(),
                    });
                }
            }
            let target = self.resolve_count(channel, count);
            self.await_count(channel, target).await?;
            captures.insert(channel.clone(), self.capture(channel, target));
        }
        Ok(captures)
    }

    fn resolve_count(&self, channel: &str, count: &Count) -> i64 {
        match count.as_int() {
            Some(n) => n,
            None => *self
                .inner
                .borrow()
                .producers_declared
                .get(channel)
                .unwrap_or(&0),
        }
    }

    async fn await_count(&self, channel: &str, target: i64) -> Result<(), StoreError> {
        if target <= 0 {
            return Ok(());
        }
        loop {
            // Check-then-park. Single-threaded: nothing runs between this
            // synchronous check and the `.notified().await` below, so a
            // `notify_waiters` from a concurrent writer can't be lost.
            let notify = {
                let inner = self.inner.borrow();
                if *inner.arrival_count.get(channel).unwrap_or(&0) >= target {
                    return Ok(());
                }
                if let Some(reason) = inner.orphaned.get(channel) {
                    return Err(StoreError::Orphaned {
                        channel: channel.to_string(),
                        reason: reason.clone(),
                    });
                }
                let arrival = *inner.arrival_count.get(channel).unwrap_or(&0);
                let remaining = *inner.producers_remaining.get(channel).unwrap_or(&0);
                // This reader's count can no longer be met: orphan THIS reader
                // only (a lower-count reader on the same channel may still be
                // satisfiable, so the channel itself is not poisoned here).
                if arrival + remaining < target {
                    return Err(StoreError::Orphaned {
                        channel: channel.to_string(),
                        reason: "insufficient_producers_remaining".to_string(),
                    });
                }
                inner
                    .notifiers
                    .get(channel)
                    .cloned()
                    .expect("channel has a notifier")
            };
            notify.notified().await;
        }
    }

    fn capture(&self, channel: &str, target: i64) -> VersionCapture {
        if target <= 0 {
            return VersionCapture::default();
        }
        let inner = self.inner.borrow();
        let entries = &inner.logs[channel];
        let mut non_init: Vec<&LogEntry> = entries.iter().filter(|e| e.write_seq != 0).collect();
        non_init.sort_by(|a, b| entry_order(a, b));
        let chosen = non_init.iter().take(target as usize);
        VersionCapture {
            captured_seqs: chosen.map(|e| e.write_seq).collect(),
        }
    }

    /// Wake every reader parked on `channel`; each re-checks its own count and
    /// orphan state and either proceeds or re-parks.
    fn notify_channel(&self, channel: &str) {
        let notify = self.inner.borrow().notifiers.get(channel).cloned();
        if let Some(n) = notify {
            n.notify_waiters();
        }
    }

    // ---- read -----------------------------------------------------------

    /// Return the reduced value per channel at the captured versions.
    pub fn read(
        &self,
        requirements: &[(String, Count)],
        capture: &BTreeMap<String, VersionCapture>,
    ) -> Result<BTreeMap<String, ChanVal>, StoreError> {
        let inner = self.inner.borrow();
        let mut out = BTreeMap::new();
        for (channel, _count) in requirements {
            if !inner.specs.contains_key(channel) {
                return Err(StoreError::UnknownChannel(channel.clone()));
            }
            let cap = capture.get(channel).ok_or_else(|| {
                StoreError::UnknownChannel(format!("no capture for channel {channel:?}"))
            })?;
            out.insert(
                channel.clone(),
                reduce_value_channel(&inner, channel, &cap.captured_seqs)?,
            );
        }
        Ok(out)
    }

    // ---- snapshots ------------------------------------------------------

    /// Final user-visible view of every channel.
    pub fn snapshot(&self) -> Result<BTreeMap<String, ChanVal>, StoreError> {
        let inner = self.inner.borrow();
        let mut out = BTreeMap::new();
        for ch in inner.specs.keys() {
            let entries = &inner.logs[ch];
            if entries.is_empty() {
                out.insert(ch.clone(), ChanVal::Unset);
                continue;
            }
            out.insert(ch.clone(), reduce_all(&inner, ch, entries)?);
        }
        Ok(out)
    }

    /// Latest committed write sequence number (0 when only init seeds exist).
    pub fn current_seq(&self) -> i64 {
        self.inner.borrow().last_seq
    }

    /// Reduce every channel considering only entries with `write_seq <= max_seq`.
    /// Unset channels are omitted from the result (absent == unwritten).
    pub fn snapshot_at_seq(&self, max_seq: i64) -> Result<BTreeMap<String, ChanVal>, StoreError> {
        let inner = self.inner.borrow();
        let mut out = BTreeMap::new();
        for ch in inner.specs.keys() {
            let entries: Vec<&LogEntry> = inner.logs[ch]
                .iter()
                .filter(|e| e.write_seq <= max_seq)
                .collect();
            if entries.is_empty() {
                continue;
            }
            let owned: Vec<LogEntry> = entries.into_iter().cloned().collect();
            out.insert(ch.clone(), reduce_all(&inner, ch, &owned)?);
        }
        Ok(out)
    }

    // ---- producer accounting / orphan propagation ----------------------

    /// Inform the store that one producer of `channel` has terminated.
    pub fn mark_producer_done(&self, channel: &str, success: bool) -> Result<(), StoreError> {
        {
            let mut inner = self.inner.borrow_mut();
            if !inner.specs.contains_key(channel) {
                return Err(StoreError::UnknownChannel(channel.to_string()));
            }
            if *inner.producers_remaining.get(channel).unwrap_or(&0) > 0 {
                *inner.producers_remaining.get_mut(channel).unwrap() -= 1;
            }
            let arrival = *inner.arrival_count.get(channel).unwrap_or(&0);
            let remaining = *inner.producers_remaining.get(channel).unwrap_or(&0);
            let has_init_seed = inner
                .logs
                .get(channel)
                .map(|es| es.iter().any(|e| e.write_seq == 0))
                .unwrap_or(false);
            // The channel is truly dead — no producer will ever write, nothing
            // was written, no init seed — so poison it: every reader errors.
            if !success && remaining == 0 && arrival == 0 && !has_init_seed {
                inner
                    .orphaned
                    .insert(channel.to_string(), "all_producers_cancelled".to_string());
            }
        }
        // Wake parked readers: an over-target reader re-checks reachability and
        // orphans itself; a lower-count reader re-parks or proceeds.
        self.notify_channel(channel);
        Ok(())
    }
}

fn validate_write_channel(
    inner: &StoreInner,
    channel: &str,
    writer_node_id: &str,
) -> Result<(), StoreError> {
    let spec = inner
        .specs
        .get(channel)
        .ok_or_else(|| StoreError::UnknownChannel(channel.to_string()))?;
    if spec.reducer != ReducerName::Overwrite {
        return Ok(());
    }
    if let Some(prior) = inner.overwrite_writer.get(channel) {
        return Err(StoreError::Reducer(ReducerError::OverwriteConflict(
            format!("{prior} already; rejecting second writer {writer_node_id}"),
        )));
    }
    Ok(())
}

fn commit_write_channel(
    inner: &mut StoreInner,
    channel: &str,
    value: &Value,
    writer_node_id: &str,
) {
    inner.last_seq += 1;
    let seq = inner.last_seq;
    inner.logs.get_mut(channel).unwrap().push(LogEntry {
        write_seq: seq,
        writer_node_id: writer_node_id.to_string(),
        value: value.clone(),
    });
    *inner.arrival_count.get_mut(channel).unwrap() += 1;
    if inner.specs[channel].reducer == ReducerName::Overwrite {
        inner
            .overwrite_writer
            .entry(channel.to_string())
            .or_insert_with(|| writer_node_id.to_string());
    }
}

fn reduce_value_channel(
    inner: &StoreInner,
    channel: &str,
    captured_seqs: &[i64],
) -> Result<ChanVal, StoreError> {
    let spec = &inner.specs[channel];
    let entries = &inner.logs[channel];
    let init = entries.iter().find(|e| e.write_seq == 0);
    let current = match init {
        Some(e) => ChanVal::Val(e.value.clone()),
        None => ChanVal::Unset,
    };
    if captured_seqs.is_empty() {
        return Ok(current);
    }
    let seq_set: std::collections::HashSet<i64> = captured_seqs.iter().copied().collect();
    let mut chosen: Vec<&LogEntry> = entries
        .iter()
        .filter(|e| seq_set.contains(&e.write_seq))
        .collect();
    chosen.sort_by(|a, b| entry_order(a, b));
    let tuples: Vec<(String, Value)> = chosen
        .iter()
        .map(|e| (e.writer_node_id.clone(), e.value.clone()))
        .collect();
    Ok(apply_reducer(spec.reducer, &current, &tuples)?)
}

fn reduce_all(
    inner: &StoreInner,
    channel: &str,
    entries: &[LogEntry],
) -> Result<ChanVal, StoreError> {
    let mut sorted: Vec<&LogEntry> = entries.iter().collect();
    sorted.sort_by(|a, b| entry_order(a, b));
    let init = sorted.iter().find(|e| e.write_seq == 0);
    let current = match init {
        Some(e) => ChanVal::Val(e.value.clone()),
        None => ChanVal::Unset,
    };
    let tuples: Vec<(String, Value)> = sorted
        .iter()
        .filter(|e| e.write_seq != 0)
        .map(|e| (e.writer_node_id.clone(), e.value.clone()))
        .collect();
    let spec = &inner.specs[channel];
    if tuples.is_empty() {
        Ok(current)
    } else {
        Ok(apply_reducer(spec.reducer, &current, &tuples)?)
    }
}

File: src/aiperf/graph/channel_store.py
# SPDX-FileCopyrightText: Copyright (c) 2025-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
"""VersionedChannelStore - per-trace channel state for the async-dataflow executor.

A versioned log per channel. Writes are linearized via a single monotonic
`_next_seq` counter; readers capture the per-channel versions at firing time and
reducers consume them in (write_seq, writer_node_id) order on read. No deepcopy
anywhere - values are stored share-by-ref.
"""

from __future__ import annotations

import asyncio
import contextlib
from collections.abc import Callable
from dataclasses import dataclass, field
from typing import Any

from aiperf.dataset.graph.models import ChannelRequirement, ChannelSpec, ReducerName
from aiperf.graph.reducers import (
    UNSET,
    OverwriteConflictError,
    Reducer,
    get_reducer,
)

__all__ = [
    "ChannelCaptureMissingError",
    "ChannelOrphanedError",
    "ReducerRegistry",
    "UnknownChannelError",
    "VersionedChannelStore",
]


class UnknownChannelError(KeyError):
    """Raised when writing or reading a channel not declared in `channel_specs`."""


class ChannelCaptureMissingError(KeyError):
    """Raised when a read's requirement set and its version capture disagree.

    Distinct from `UnknownChannelError`: the channel IS declared, it just has
    no entry in the capture the caller passed to `read`. Both are built by the
    executor from the same firing, so a mismatch is an executor invariant
    violation, not a missing write and not bad workload input. `KeyError` is
    kept as the base so any caller catching the broad type still catches this.
    """


class ChannelOrphanedError(RuntimeError):
    """Raised when an `await_inputs` requirement can no longer be satisfied.

    Surfaces when a channel's arrivals-so-far plus its still-live producers can
    no longer reach the waiter's count target -- which can happen while
    producers are still running, not only after every producer is done.
    """


# Type alias for the reducer dispatch dependency. The store does not depend
# on a concrete registry class - any callable that maps reducer-name to a
# `(current, [(writer_id, value), ...]) -> reduced` reducer works. Defaults
# to `aiperf.graph.reducers.get_reducer`.
ReducerRegistry = Callable[[ReducerName], Reducer]


@dataclass(slots=True, frozen=True)
class _LogEntry:
    """One committed write to a value channel."""

    write_seq: int
    """Monotonic global sequence number assigned at commit time."""
    writer_node_id: str
    """Identifier of the node that produced this write; secondary sort key."""
    value: Any
    """The raw write value; stored by reference, never deep-copied."""


@dataclass(slots=True, frozen=True)
class _VersionCapture:
    """Immutable snapshot of which channel versions count for one firing."""

    channel: str
    """Channel this capture refers to."""
    captured_seqs: tuple[int, ...]
    """`write_seq` values included in this capture, in commit order."""


@dataclass(slots=True)
class _Waiter:
    """One pending `await_inputs` requirement on a single channel."""

    channel: str
    required_count: int
    event: asyncio.Event = field(default_factory=asyncio.Event)
    orphaned_reason: str | None = None


class VersionedChannelStore:
    """Per-trace channel store backed by per-channel append-only logs.

    Writers call `write(channels, value, writer_node_id=...)`; readers call
    `await await_inputs(reqs)` to capture versions, then `read(reqs, capture)`
    to materialize reduced values. Cancellation is communicated by the
    executor via `mark_producer_done(channel, success=...)`.
    """

    __slots__ = (
        "_specs",
        "_logs",
        "_arrival_count",
        "_producers_remaining",
        "_producers_declared",
        "_initial",
        "_next_seq",
        "_waiters",
        "_reducers",
        "_orphaned",
        "_overwrite_writer",
    )

    def __init__(
        self,
        initial: dict[str, Any],
        channel_specs: dict[str, ChannelSpec],
        producers_per_channel: dict[str, int],
    ) -> None:
        self._specs: dict[str, ChannelSpec] = dict(channel_specs)
        self._logs: dict[str, list[_LogEntry]] = {ch: [] for ch in channel_specs}
        self._arrival_count: dict[str, int] = dict.fromkeys(channel_specs, 0)
        self._producers_remaining: dict[str, int] = {
            ch: int(producers_per_channel.get(ch, 0)) for ch in channel_specs
        }
        self._producers_declared: dict[str, int] = {
            ch: int(producers_per_channel.get(ch, 0)) for ch in channel_specs
        }
        self._initial: dict[str, Any] = dict(initial)
        # `_next_seq` is incremented before every commit. Reserve 0 for
        # synthetic initial-state writes; node writes start at 1.
        self._next_seq: int = 0
        self._waiters: dict[str, list[_Waiter]] = {ch: [] for ch in channel_specs}
        self._reducers: ReducerRegistry = get_reducer
        self._orphaned: dict[str, str] = {}
        # Tracks the writer id of the first non-init write to each overwrite
        # channel so a different writer can be rejected with
        # OverwriteConflictError across the trace's lifetime (cross-spec
        # decision 1: strict over the whole trace, not per-dispatch).
        self._overwrite_writer: dict[str, str] = {}

        for ch, value in self._initial.items():
            if ch not in self._specs:
                raise UnknownChannelError(
                    f"initial state names unknown channel: {ch!r}"
                )
            self._logs[ch].append(
                _LogEntry(write_seq=0, writer_node_id="__init__", value=value)
            )
            # Note: init seed does NOT bump `_arrival_count`. arrival_count
            # counts node writes only, so `count=N` requirements measure
            # producer arrivals.

    # ------------------------------------------------------------------
    # write
    # ------------------------------------------------------------------
    def write(
        self,
        channel_names: list[str],
        value: Any,
        *,
        writer_node_id: str,
    ) -> None:
        """Commit `value` to every channel in `channel_names`."""
        if not channel_names:
            return
        for ch in channel_names:
            self._validate_write_channel(ch, writer_node_id=writer_node_id)

        for ch in channel_names:
            self._commit_write_channel(ch, value, writer_node_id=writer_node_id)
            self._wake_waiters(ch)

    def _validate_write_channel(self, channel: str, *, writer_node_id: str) -> None:
        if channel not in self._specs:
            raise UnknownChannelError(f"unknown channel: {channel!r}")
        if self._specs[channel].reducer is not ReducerName.OVERWRITE:
            return
        prior = self._overwrite_writer.get(channel)
        if prior is None:
            return
        # Both cases are errors -- the executor fires each node at most once per
        # instance, so even a repeat write from the SAME node is an invariant
        # violation -- but they are different bugs and must not share a message.
        # The second-writer wording named one node as both the incumbent and the
        # interloper ("already written by 'n1'; rejecting second writer 'n1'"),
        # sending the reader hunting for a writer that does not exist.
        if prior == writer_node_id:
            raise OverwriteConflictError(
                f"overwrite channel {channel!r} written twice by the same node "
                f"{writer_node_id!r}: each node fires at most once per instance "
                "run, so this is an executor invariant violation, not a "
                "multi-writer conflict"
            )
        raise OverwriteConflictError(
            f"overwrite channel {channel!r} already written by "
            f"{prior!r}; rejecting second writer {writer_node_id!r}"
        )

    def has_overwrite_writer(self, channel: str) -> bool:
        """Return True when `channel` is OVERWRITE-reduced and already written.

        A subsequent `write` to such a channel raises `OverwriteConflictError`.
        Callers that must not raise (the executor's dispatch-failure containment
        path, which writes a content-neutral sentinel to the failed node's own
        output channels) check this first and skip the write instead.
        """
        if channel not in self._specs:
            raise UnknownChannelError(f"unknown channel: {channel!r}")
        return (
            self._specs[channel].reducer is ReducerName.OVERWRITE
            and channel in self._overwrite_writer
        )

    def _commit_write_channel(
        self, channel: str, value: Any, *, writer_node_id: str
    ) -> None:
        self._next_seq += 1
        entry = _LogEntry(
            write_seq=self._next_seq,
            writer_node_id=writer_node_id,
            value=value,
        )
        self._logs[channel].append(entry)
        self._arrival_count[channel] += 1
        if self._specs[channel].reducer is ReducerName.OVERWRITE:
            self._overwrite_writer.setdefault(channel, writer_node_id)

    # ------------------------------------------------------------------
    # await_inputs
    # ------------------------------------------------------------------
    async def await_inputs(
        self, requirements: list[ChannelRequirement]
    ) -> dict[str, _VersionCapture]:
        """Block until every requirement is satisfied; return frozen captures.

        For each requirement: `count=N` waits for the N-th arrival; `count="all"`
        resolves to the static topology count at call time and then behaves
        identically. Raises `ChannelOrphanedError` if any required channel can
        no longer fulfil its count.
        """
        captures: dict[str, _VersionCapture] = {}
        for req in requirements:
            ch = req.channel
            if ch not in self._specs:
                raise UnknownChannelError(f"unknown channel: {ch!r}")
            if ch in self._orphaned:
                raise ChannelOrphanedError(
                    f"channel {ch!r} orphaned: {self._orphaned[ch]}"
                )
            target = self._resolve_count(req)
            await self._await_count(ch, target)
            captures[ch] = self._capture(ch, target)
        return captures

    def _resolve_count(self, req: ChannelRequirement) -> int:
        if req.count == "all":
            # "all" resolves to the static topology count at call time. Late
            # producer cancellation may render this unreachable; orphan check
            # inside _await_count handles that case.
            return self._producers_declared[req.channel]
        return int(req.count)

    async def _await_count(self, channel: str, target: int) -> None:
        if target <= 0:
            return
        while self._arrival_count[channel] < target:
            if channel in self._orphaned:
                raise ChannelOrphanedError(
                    f"channel {channel!r} orphaned: {self._orphaned[channel]}"
                )
            # Check reachability: arrivals so far + still-live producers must
            # reach target, else orphan immediately.
            reachable = (
                self._arrival_count[channel] + self._producers_remaining[channel]
            )
            if reachable < target:
                self._orphaned[channel] = "insufficient_producers_remaining"
                raise ChannelOrphanedError(
                    f"channel {channel!r} cannot reach count={target}: "
                    f"only {reachable} arrivals possible "
                    f"({self._arrival_count[channel]} so far, "
                    f"{self._producers_remaining[channel]} producers remain)"
                )
            waiter = _Waiter(channel=channel, required_count=target)
            self._waiters[channel].append(waiter)
            try:
                await waiter.event.wait()
            finally:
                with contextlib.suppress(ValueError):
                    self._waiters[channel].remove(waiter)
            if waiter.orphaned_reason is not None:
                raise ChannelOrphanedError(
                    f"channel {channel!r} orphaned: {waiter.orphaned_reason}"
                )

    def _capture(self, channel: str, target: int) -> _VersionCapture:
        entries = self._logs[channel]
        if target <= 0:
            return _VersionCapture(channel=channel, captured_seqs=())
        # Init seed (write_seq=0) is the reducer seed, not an "arrival"; the
        # count-N requirement measures node writes only. The init entry is
        # always implicitly part of read (handled in _reduce_value_channel).
        non_init = [e for e in entries if e.write_seq != 0]
        sorted_entries = sorted(non_init, key=lambda e: (e.write_seq, e.writer_node_id))
        chosen = sorted_entries[:target]
        return _VersionCapture(
            channel=channel,
            captured_seqs=tuple(e.write_seq for e in chosen),
        )

    def _wake_waiters(self, channel: str) -> None:
        if not self._waiters[channel]:
            return
        current = self._arrival_count[channel]
        for waiter in list(self._waiters[channel]):
            if current >= waiter.required_count:
                waiter.event.set()

    # ------------------------------------------------------------------
    # read
    # ------------------------------------------------------------------
    def read(
        self,
        requirements: list[ChannelRequirement],
        capture: dict[str, _VersionCapture],
    ) -> dict[str, Any]:
        """Return the reduced value per channel at the captured versions."""
        out: dict[str, Any] = {}
        for req in requirements:
            ch = req.channel
            if ch not in self._specs:
                raise UnknownChannelError(f"unknown channel: {ch!r}")
            cap = capture.get(ch)
            if cap is None:
                raise ChannelCaptureMissingError(
                    f"channel {ch!r} is required by this node but absent from "
                    "its version capture: the requirement set and the capture "
                    "were built from different reads, which is an executor "
                    "invariant violation, not a missing write"
                )
            out[ch] = self._reduce_value_channel(ch, cap.captured_seqs)
        return out

    def _reduce(self, channel: str, chosen: list[_LogEntry]) -> Any:
        """Seed with the init write, order `chosen`, and apply the reducer.

        The single reduce sequence shared by both read paths. `chosen` holds the
        non-init entries taking part in this reduction: the captured subset for
        `read`, every node write for `snapshot`. Everything else -- which entry
        seeds the reducer, the `(write_seq, writer_node_id)` order, the
        no-writes shortcut -- is identical between the two by construction.
        """
        entries = self._logs[channel]
        # Init seed (write_seq=0) is the reducer's starting value; it is not
        # itself a participating write. Capture sets hold node writes only.
        init_entry = next((e for e in entries if e.write_seq == 0), None)
        current: Any = init_entry.value if init_entry is not None else UNSET
        if not chosen:
            return current
        ordered = sorted(chosen, key=lambda e: (e.write_seq, e.writer_node_id))
        reducer = self._reducers(self._specs[channel].reducer)
        try:
            return reducer(current, [(e.writer_node_id, e.value) for e in ordered])
        except (OverwriteConflictError, TypeError) as e:
            # Reducers are channel-agnostic by signature, so their messages
            # carry writer node ids only. The channel is the one fact that
            # localizes the failure in a graph with dozens of channels and a
            # shared writer set. Re-raised as the same class so callers
            # catching OverwriteConflictError / TypeError are unaffected.
            raise type(e)(f"channel {channel!r}: {e}") from e

    def _reduce_value_channel(
        self, channel: str, captured_seqs: tuple[int, ...]
    ) -> Any:
        seq_set = set(captured_seqs)
        return self._reduce(
            channel, [e for e in self._logs[channel] if e.write_seq in seq_set]
        )

    # ------------------------------------------------------------------
    # snapshot
    # ------------------------------------------------------------------
    def snapshot(self) -> dict[str, Any]:
        """Return the final user-visible view of every channel."""
        return {
            ch: self._reduce(ch, [e for e in self._logs[ch] if e.write_seq != 0])
            for ch in self._specs
        }

    # ------------------------------------------------------------------
    # producer accounting / orphan propagation
    # ------------------------------------------------------------------
    def mark_producer_done(
        self,
        channel: str,
        *,
        success: bool,
    ) -> None:
        """Inform the store that one producer of `channel` has terminated.

        Decrements the remaining-producer count, then -- regardless of
        `success` -- wakes every waiter whose count target is now unreachable
        (arrivals + remaining producers), stamping `orphaned_reason`. `success`
        only picks the label: `insufficient_producers_remaining` vs
        `all_producers_cancelled`.

        The CHANNEL itself is marked orphaned (failing later `await_inputs`
        immediately) only when `success=False` leaves it with no producers
        remaining, no arrivals, and no init-state seed. Whether this particular
        producer wrote is irrelevant: arrivals are already counted by `write`.
        """
        if channel not in self._specs:
            raise UnknownChannelError(f"unknown channel: {channel!r}")
        if self._producers_remaining[channel] > 0:
            self._producers_remaining[channel] -= 1
        reachable = self._arrival_count[channel] + self._producers_remaining[channel]
        orphan_reason = (
            "insufficient_producers_remaining" if success else "all_producers_cancelled"
        )
        for waiter in list(self._waiters[channel]):
            if reachable < waiter.required_count:
                waiter.orphaned_reason = orphan_reason
                waiter.event.set()
        has_init_seed = channel in self._logs and any(
            entry.write_seq == 0 for entry in self._logs[channel]
        )
        if (
            not success
            and self._producers_remaining[channel] == 0
            and self._arrival_count[channel] == 0
            and not has_init_seed
        ):
            self._orphaned[channel] = orphan_reason
